### PR TITLE
Add container_perf_event_error_code metric

### PR DESCRIPTION
**Pull request by kwisniewski98**
Signed-off-by: Wisniewski, Krzysztof2 <Krzysztof2.Wisniewski@intel.com>

### CODE REVIEW

File: docs/storage/prometheus.md [hunk 1]
```diff
@@ -90,6 +90,7 @@ Metric name | Type | Description | Unit (where applicable) | -disable_metrics pa
 `container_spec_memory_reservation_limit_bytes` | Gauge | Memory reservation limit for the container | bytes | |
 `container_start_time_seconds` | Gauge | Start time of the container since unix epoch | seconds | |
 `container_tasks_state` | Gauge | Number of tasks in given state (`sleeping`, `running`, `stopped`, `uninterruptible`, or `ioawaiting`) | | |
+`container_perf_event_errors` | Gaouge | Error code returned by opening or decoding perf event | | | libpfm
```
Comment (wacuuu) [inline docs/storage/prometheus.md]: Typo in Gaouge -> Gauge

File: metrics/testdata/prometheus_metrics_perf_aggregated [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # HELP cadvisor_version_info A metric with a constant '1' value labeled by kernel version, OS version, docker version, cadvisor version & cadvisor revision.
 # TYPE cadvisor_version_info gauge
 cadvisor_version_info{cadvisorRevision="abcdef",cadvisorVersion="0.16.0",dockerVersion="1.8.1",kernelVersion="4.1.6-200.fc22.x86_64",osVersion="Fedora 22 (Twenty Two)"} 1
+ # HELP container_perf_event_errors Perf event errors.
+        # TYPE container_perf_event_errors gauge
```
Comment (wacuuu) [inline metrics/testdata/prometheus_metrics_perf_aggregated]: This indentation seems wrong?

File: info/v2/conversion.go [hunk 3]
```diff
@@ -152,11 +152,11 @@ func ContainerStatsFromV1(containerName string, spec *v1.ContainerSpec, stats []
...
 		}
-		if len(val.PerfStats) > 0 {
-			stat.PerfStats = val.PerfStats
+		if len(val.Perf.PerfStats) > 0 {
+			stat.Perf.PerfStats = val.Perf.PerfStats
 		}
-		if len(val.PerfUncoreStats) > 0 {
-			stat.PerfUncoreStats = val.PerfUncoreStats
+		if len(val.Perf.PerfUncoreStats) > 0 {
+			stat.Perf.PerfUncoreStats = val.Perf.PerfUncoreStats
```
Comment (katarzyna-z) [inline info/v2/conversion.go:159]: You've added `PerfErrors` and it seems that these elements should be also available here (and available on API v2). Please test it and use appropriate form.

File: info/v2/conversion.go [hunk 4]
```diff
@@ -213,11 +213,11 @@ func DeprecatedStatsFromV1(cont *v1.ContainerInfo) []DeprecatedContainerStats {
 		if stat.HasCustomMetrics {
 			stat.CustomMetrics = val.CustomMetrics
 		}
-		if len(val.PerfStats) > 0 {
-			stat.PerfStats = val.PerfStats
+		if len(val.Perf.PerfStats) > 0 {
+			stat.Perf.PerfStats = val.Perf.PerfStats
```
Comment (katarzyna-z) [inline info/v2/conversion.go:220]: the same as above

File: info/v2/conversion_test.go [hunk 5]
```diff
@@ -197,40 +197,42 @@ func TestContainerStatsFromV1(t *testing.T) {
...
-					Value:        654321,
-					Name:         "uncore_imc_0/cas_count_write",
+				{
+					PerfValue: v1.PerfValue{
+						ScalingRatio: 1.0,
+						Value:        654321,
+						Name:         "uncore_imc_0/cas_count_write",
+					},
+					Socket: 1,
+					PMU:    "17",
```
Comment (katarzyna-z) [inline info/v2/conversion_test.go:234]: This test should be extended to cover also `PerfErrors`.

File: info/v2/conversion_test.go [hunk 6]
```diff
@@ -270,9 +272,11 @@ func TestContainerStatsFromV1(t *testing.T) {
 			BaseUsageBytes:  &v1Stats.Filesystem[0].BaseUsage,
 			InodeUsage:      &v1Stats.Filesystem[0].Inodes,
 		},
-		Accelerators:     v1Stats.Accelerators,
-		PerfStats:        v1Stats.PerfStats,
-		PerfUncoreStats:  v1Stats.PerfUncoreStats,
+		Accelerators: v1Stats.Accelerators,
+		Perf: v1.Perf{
+			PerfStats:       v1Stats.Perf.PerfStats,
+			PerfUncoreStats: v1Stats.Perf.PerfUncoreStats,
```
Comment (katarzyna-z) [inline info/v2/conversion_test.go:285]: Imho you should also add here `PerfErrors`.

File: cmd/internal/storage/statsd/statsd.go [hunk 7]
```diff
@@ -178,7 +178,7 @@ func (s *statsdStorage) hugetlbStatsToValues(series *map[string]uint64, stats *i
 }
 
 func (s *statsdStorage) perfStatsToValues(series *map[string]uint64, stats *info.ContainerStats) {
-	for _, perfStat := range stats.PerfStats {
+	for _, perfStat := range stats.Perf.PerfStats {
```
Comment (katarzyna-z) [inline cmd/internal/storage/statsd/statsd.go:181]: Have you tests manually integration with statsd as unit tests are not available for this integration?

File: cmd/internal/storage/stdout/stdout.go [hunk 8]
```diff
@@ -183,7 +183,7 @@ func (driver *stdoutStorage) hugetlbStatsToValues(series *map[string]uint64, sta
 }
 
 func (driver *stdoutStorage) perfStatsToValues(series *map[string]uint64, stats *info.ContainerStats) {
-	for _, perfStat := range stats.PerfStats {
+	for _, perfStat := range stats.Perf.PerfStats {
```
Comment (katarzyna-z) [inline cmd/internal/storage/stdout/stdout.go:186]: Have you tests manually with stdout strorage as unit tests are not available?

File: info/v1/container.go [hunk 9]
```diff
@@ -921,6 +921,22 @@ type ProcessStats struct {
...
+
+type Perf struct {
+	// Statistics originating from perf events
+	PerfStats []PerfStat `json:"perf_stats,omitempty"`
+
+	// Statistics originating from perf uncore events.
+	// Applies only for root container.
+	PerfUncoreStats []PerfUncoreStat `json:"perf_uncore_stats,omitempty"`
+
+	PerfErrors []PerfError `json:"perf_errors"`
```
Comment (katarzyna-z) [inline info/v1/container.go:940]: Please add description in the code for this elements as it is for PerfStat and PerfUncoreStats.

File: info/v1/container.go [hunk 10]
```diff
@@ -921,6 +921,22 @@ type ProcessStats struct {
 	// Ulimits for the top-level container process
 	Ulimits []UlimitSpec `json:"ulimits,omitempty"`
 }
+type PerfError struct {
+	EventName string `json:"event_name"`
+	Action    string `json:"action"`
```
Comment (katarzyna-z) [inline info/v1/container.go:927]: What does "Action" stand for? Could you add descriptions for elements in PerfError?

File: perf/collector_libpfm.go [hunk 11]
```diff
@@ -184,35 +193,20 @@ func (c *collector) setup() error {
 	c.cpuFilesLock.Lock()
 	defer c.cpuFilesLock.Unlock()
 	cgroupFd := int(cgroup.Fd())
-	for i, group := range c.events.Core.Events {
+	i := 0
```
Comment (katarzyna-z) [inline perf/collector_libpfm.go]: It seems that `i` variable has a specific meaning in this case so please change the name of this variable to more meaningful.

File: perf/collector_libpfm.go [hunk 12]
```diff
@@ -271,11 +294,17 @@ func (c *collector) registerEvent(event eventInfo, leaderFileDescriptors map[int
 	for _, cpu := range c.onlineCPUs {
 		fd, err := unix.PerfEventOpen(event.config, pid, cpu, leaderFileDescriptors[cpu], flags)
 		if err != nil {
-			return nil, fmt.Errorf("setting up perf event %#v failed: %q", event.config, err)
+			errorCode, _ := strconv.Atoi(fmt.Sprintf("%s", err))
```
Comment (katarzyna-z) [inline perf/collector_libpfm.go]: Please handle an error for strconv.Atoi.

File: metrics/testdata/prometheus_metrics [hunk 13]
```diff
@@ -341,6 +341,9 @@ container_network_udp_usage_total{container_env_foo_env="prod",container_label_f
 container_network_udp_usage_total{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",udp_state="listen",zone_name="hello"} 0 1395066363000
 container_network_udp_usage_total{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",udp_state="rxqueued",zone_name="hello"} 0 1395066363000
 container_network_udp_usage_total{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",udp_state="txqueued",zone_name="hello"} 0 1395066363000
+# HELP container_perf_event_errors Perf event errors.
+# TYPE container_perf_event_errors gauge
+container_perf_event_errors{action="some_action",container_env_foo_env="prod",container_label_foo_label="bar",event="non_existing_event",id="testcontainer",image="test",name="testcontaineralias",zone_name="hello"} -4 1395066363000
```
Comment (katarzyna-z) [inline metrics/testdata/prometheus_metrics]: In really collection if there is an error we will have multiple errors for many CPUs or many PMUs and sockets in this format of metric we will have many metrics with the same labels and the same value. Have you checked how it works with Prometheus?
Please add more complicated case to unit tests.

**Review by katarzyna-z — CHANGES_REQUESTED** [review]

**Comment by katarzyna-z** [comment]
@Creatone please take a look at this pull request.

File: cmd/internal/storage/stdout/stdout.go (hunk 8 above)
Comment (kwisniewski98) [reply cmd/internal/storage/stdout/stdout.go:186]: Seems to be working fine

File: cmd/internal/storage/statsd/statsd.go (hunk 7 above)
Comment (kwisniewski98) [reply cmd/internal/storage/statsd/statsd.go:181]: Seems to be working fine

File: docs/storage/prometheus.md [hunk 14]
```diff
@@ -90,6 +90,7 @@ Metric name | Type | Description | Unit (where applicable) | -disable_metrics pa
 `container_spec_memory_reservation_limit_bytes` | Gauge | Memory reservation limit for the container | bytes | |
 `container_start_time_seconds` | Gauge | Start time of the container since unix epoch | seconds | |
 `container_tasks_state` | Gauge | Number of tasks in given state (`sleeping`, `running`, `stopped`, `uninterruptible`, or `ioawaiting`) | | |
+`container_perf_event_errors` | Gauge | Error code returned by opening or decoding perf event | | | libpfm
```
Comment (Creatone) [inline docs/storage/prometheus.md]: Should be `container_perf_event_error` ?

File: metrics/prometheus.go [hunk 15]
```diff
@@ -1653,6 +1653,26 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 					return values
 				},
 			},
+			{
+				name:        "container_perf_event_errors",
+				help:        "Perf event errors.",
```
Comment (Creatone) [inline metrics/prometheus.go]: Should be `Perf event error code`?

File: metrics/prometheus.go [hunk 16]
```diff
@@ -1653,6 +1653,26 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 					return values
 				},
 			},
+			{
+				name:        "container_perf_event_errors",
```
Comment (Creatone) [inline metrics/prometheus.go]: Similar to comment from `docs/storage/prometheus.md`

File: perf/collector_libpfm.go [hunk 17]
```diff
@@ -231,7 +225,36 @@ func (c *collector) setup() error {
...
+	for j, event := range events {
+		// First element is group leader.
+		isGroupLeader := j == 0
+		customEvent, ok := c.eventToCustomEvent[event]
+		var err error
+		if ok {
+			config := c.createConfigFromRawEvent(customEvent)
+			leaderFileDescriptors, err = c.registerEvent(eventInfo{string(customEvent.Name), config, cgroupFd, groupIndex, isGroupLeader}, leaderFileDescriptors)
+			if err != nil {
+				return nil, fmt.Errorf("Unable to register perf event %v", err)
```
Comment (Creatone) [inline perf/collector_libpfm.go]: Error string shouldn't be capitalized.

File: perf/collector_libpfm.go [hunk 18]
```diff
@@ -231,7 +225,36 @@ func (c *collector) setup() error {
...
+			}
+		} else {
+			config, err := c.createConfigFromEvent(event)
+			if err != nil {
+				return nil, fmt.Errorf("Unable to create config from perf event %v", err)
```
Comment (Creatone) [inline perf/collector_libpfm.go]: Error string shouldn't be capitalized.

File: perf/collector_libpfm.go [hunk 19]
```diff
@@ -231,7 +225,36 @@ func (c *collector) setup() error {
...
+
+			}
+			leaderFileDescriptors, err = c.registerEvent(eventInfo{string(event), config, cgroupFd, groupIndex, isGroupLeader}, leaderFileDescriptors)
+			if err != nil {
+				return nil, fmt.Errorf("Unable to register perf event %v", err)
```
Comment (Creatone) [inline perf/collector_libpfm.go]: Error string shouldn't be capitalized.

File: perf/collector_libpfm.go [hunk 20]
```diff
@@ -265,17 +288,26 @@ func (c *collector) registerEvent(event eventInfo, leaderFileDescriptors map[int
...
 	}
-
 	setAttributes(event.config, event.isGroupLeader)
 
 	for _, cpu := range c.onlineCPUs {
+		err := fmt.Errorf("-1")
 		fd, err := unix.PerfEventOpen(event.config, pid, cpu, leaderFileDescriptors[cpu], flags)
 		if err != nil {
-			return nil, fmt.Errorf("setting up perf event %#v failed: %q", event.config, err)
+			errorCode, conversionError := strconv.Atoi(fmt.Sprintf("%s", err))
```
Comment (Creatone) [inline perf/collector_libpfm.go]: Consider using `err.Error()` instead of `fmt.Sptintf("%s", err)`

File: perf/collector_libpfm.go [hunk 21]
```diff
@@ -265,17 +288,26 @@ func (c *collector) registerEvent(event eventInfo, leaderFileDescriptors map[int
 		pid = -1
 		flags = unix.PERF_FLAG_FD_CLOEXEC
 	}
-
 	setAttributes(event.config, event.isGroupLeader)
 
 	for _, cpu := range c.onlineCPUs {
+		err := fmt.Errorf("-1")
```
Comment (Creatone) [inline perf/collector_libpfm.go]: Consider using `errors.New("-1")` instead of `fmt.Errorf("-1")`.

File: perf/uncore_libpfm.go [hunk 22]
```diff
@@ -202,7 +204,7 @@ func (c *uncoreCollector) setup(events PerfEvents, devicesPath string) error {
 			}
 
 			if err != nil {
-				return err
+				klog.Errorf("%v", err)
```
Comment (Creatone) [inline perf/uncore_libpfm.go]: Consider using `klog.Error(err)`

File: perf/uncore_libpfm.go [hunk 23]
```diff
@@ -397,6 +405,12 @@ func (c *uncoreCollector) registerEvent(eventInfo eventInfo, pmu pmu, leaderFile
 		groupFd, flags := leaderFileDescriptors[cpu], 0
 		fd, err := c.perfEventOpen(eventInfo.config, eventInfo.pid, int(cpu), groupFd, flags)
 		if err != nil {
+			errorCode, _ := strconv.Atoi(fmt.Sprintf("%s", err))
```
Comment (Creatone) [inline perf/uncore_libpfm.go]: Consider using `err.Error()` instead of `fmt.Sptintf("%s", err)`

**Review by Creatone — CHANGES_REQUESTED** [review]

File: info/v1/container.go [hunk 24]
```diff
@@ -921,6 +921,24 @@ type ProcessStats struct {
 	// Ulimits for the top-level container process
 	Ulimits []UlimitSpec `json:"ulimits,omitempty"`
 }
+type PerfError struct {
```
Comment (katarzyna-z) [inline info/v1/container.go:924]: I think that all elements in PerfError struct could have a comment to keep style of  Perf struct.

File: metrics/prometheus.go [hunk 25]
```diff
@@ -1907,6 +1907,18 @@ func (c *PrometheusCollector) collectContainersInfo(ch chan<- prometheus.Metric)
 				}
 			}
 		}
+		if c.includedMetrics.Has(container.PerfMetrics) {
+			perfErrors = append(perfErrors, cont.Stats[0].Perf.PerfErrors...)
```
Comment (katarzyna-z) [inline metrics/prometheus.go:1911]: Is is checked somehow that cont.Stats contains at least one element?

File: metrics/prometheus.go [hunk 26]
```diff
@@ -1907,6 +1907,18 @@ func (c *PrometheusCollector) collectContainersInfo(ch chan<- prometheus.Metric)
...
+		}
+	}
+	if c.includedMetrics.Has(container.PerfMetrics) {
```
Comment (katarzyna-z) [inline metrics/prometheus.go:1914]: I'm not sure if it a good place for creating this metrics, take a look what I made for machine metrics: https://github.com/google/cadvisor/blob/master/metrics/prometheus_machine.go#L129
Are now metrics unique? Where are metrics filtered to have only one metric per container?

File: perf/collector_libpfm.go [hunk 27]
```diff
@@ -184,35 +193,20 @@ func (c *collector) setup() error {
...
-					return err
-				}
-				leaderFileDescriptors, err = c.registerEvent(eventInfo{string(event), config, cgroupFd, i, isGroupLeader}, leaderFileDescriptors)
-				if err != nil {
-					return err
-				}
-				// Clean memory allocated by C code.
-				C.free(unsafe.Pointer(config))
-			}
+		leaderFileDescriptors, err := c.createLeaderFileDescriptors(group.events, cgroupFd, groupIndex, leaderFileDescriptors)
```
Comment (katarzyna-z) [inline perf/collector_libpfm.go:204]: I think that introducing of a new metric for perf errors and changes in counting perf metrics should be splitted into two separated pull requests.

Chages which are made in setup() are not related to a new metric.

File: metrics/prometheus.go (hunk 25 above)
Comment (kwisniewski98) [reply metrics/prometheus.go:1911]: It is checked in line [1878](https://github.com/kwisniewski98/cadvisor/pull/7/files#diff-88eab3cc8cef9ad727beea9923056cbbR1878)

File: perf/collector_libpfm.go [hunk 28]
```diff
@@ -184,35 +193,20 @@ func (c *collector) setup() error {
...
+		if err != nil {
+			klog.Errorf("%v", err)
```
Comment (Creatone) [inline perf/collector_libpfm.go:206]: `klog.Error(err)`

**Review by Creatone — CHANGES_REQUESTED** [review]

File: metrics/prometheus.go (hunk 26 above)
Comment (kwisniewski98) [reply metrics/prometheus.go:1914]: As far as I know this is only place for creating that metric. In prometheus_machine.go we don't have access to those structs. Metrics are now unique. It reports unique entries without any filtering from our side, so I suppose, that prometheus filters it out.

**Comment by katarzyna-z** [comment]
Could you limit the scope of this pull request only to thing which is describe in the title "Add container_perf_event_error_code metric"?

Please resolve conflicts.

**Pull request closed without merging**